autogluon/utils/tabular/features/generators/abstract.py

```diff
@@ -0,0 +1,520 @@
+        Because of how the generators track output features and types, it is generally required that the data be transformed during fit, so the fit function is rarely useful to implement beyond a simple call to fit_transform.
+
+        Parameters
+        ----------
+        X : DataFrame
+            Input data used to fit the generator.
+        **kwargs
+            Any additional arguments that a particular generator implementation could use.
+            See fit_transform method for common kwargs values.
+        """
+        self.fit_transform(X, **kwargs)
+
+    def fit_transform(self, X: DataFrame, y: Series = None, feature_metadata_in: FeatureMetadata = None, **kwargs) -> DataFrame:
+        """
+        Fit generator to the provided data and return the transformed version of the data as if fit and transform were called sequentially with the same data.
+        This is generally more efficient than calling fit and transform separately and can be up to twice as fast if the fit process requires transformation of the data.
+        This cannot be called after the generator has been fit, and will result in an AssertionError.
+
+        Parameters
+        ----------
+        X : DataFrame
+            Input data used to fit the generator.
+        y : Series, optional
+            Input data's labels used to fit the generator. Most generators do not utilize labels.
+            y.index must be equal to X.index to avoid misalignment.
+        feature_metadata_in : FeatureMetadata, optional
+            Identical to providing feature_metadata_in during generator initialization. Ignored if self.feature_metadata_in is already specified.
+            If neither are set, feature_metadata_in will be inferred from the _infer_feature_metadata_in method.
+        **kwargs
+            Any additional arguments that a particular generator implementation could use. Passed to _fit_transform and _fit_generators methods.
+
+        Returns
+        -------
+        X_out : DataFrame object which is the transformed version of the input data X.
+
+        """
+        start_time = time.time()
+        self._log(20, f'Fitting {self.__class__.__name__}...')
+        if self._is_fit:
+            raise AssertionError(f'{self.__class__.__name__} is already fit.')
+        if y is not None and isinstance(y, Series):
+            if list(y.index) != list(X.index):
+                raise AssertionError(f'y.index and X.index must be equal when fitting {self.__class__.__name__}, but they differ.')
+
+        if self.reset_index:
+            X_index = copy.deepcopy(X.index)
+            X = X.reset_index(drop=True)  # TODO: Theoretically inplace=True avoids data copy, but can lead to altering of original DataFrame outside of method context.
+            if y is not None and isinstance(y, Series):
+                y = y.reset_index(drop=True)  # TODO: this assumes y and X had matching indices prior
+        else:
+            X_index = None
```

Comment:
move into separate method to improve readability (same comment about other feature sections of this block below).

Also design consideration: the code in this block also look like a set of pre-processors: does it makes sense to consider converting those to generators (or pre-processors) too? This would allow to move out the code related to specific aspect (like columns fixing) into smaller, more readable chunk of code with it's own fit-transform methods.

Reply:
This particular code chunk can't be neatly separated nor turned into a separate generator because its stateful variable is the X_index which can only be applied after all other generators make their transforms (its logic must be applied at pre-processing, then inverse applied at post-processing).

The reason why the other preprocessor logic is not separated into a pre_generators argument akin to post_generators is because of the confusion it causes in code readability due to feature_metadata_in not being usable anymore in the actual _fit_transform logic, and some intermediate feature_metadata must be used after pre_generators.

This could theoretically be cleaned up to avoid such confusion, but the logic is complex and it is a 2-way door since the number of pre_generators that make sense to use is rather limited at present (and because BulkFeatureGenerator has pre_generators functionality where it is actually important to have).